Handle top level <style> elements automatically

dfreedm opened the pull request:
Remove the need to explicitly call into CustomStyleInterface

Related to #134

dfreedm committed:
Handle top level <style> elements automatically

Remove the need to explicitly call into CustomStyleInterface

Related to #134

azakus commented:
Hmm, weird error in edge with `getRootNode` failing.

dfreedm committed:
fix async-loading test and update dependencies

dfreedm committed:
make sure template polyfill is in async loading test

sorvell commented on entrypoints/apply-shim.js:
```diff
@@ -36,18 +36,21 @@ class ApplyShimInterface {
-        });
-      }
+    if (!this.customStyleInterface) {
+      this.customStyleInterface = new CustomStyleInterface();
```
Seems like it might make sense to have an import that can bring in this singleton instance rather than this slightly tricky conditional here?

azakus replied:
Yep, will do!

sorvell commented on entrypoints/apply-shim.js:
```diff
@@ -36,18 +36,21 @@ class ApplyShimInterface {
+      this.customStyleInterface = new CustomStyleInterface();
+      window.ShadyCSS.CustomStyleInterface = this.customStyleInterface;
     }
+    this.customStyleInterface['transformCallback'] = (style) => {
```
Do these hooks not have to compose? If so, maybe there could be an add/remove hook?

azakus replied:
No, the ScopingShim's version of these hooks entirely supersede ApplyShim's version, so there is no need to compose them.

sorvell commented on src/custom-style-interface.js:
```diff
@@ -105,11 +107,51 @@ export default class CustomStyleInterface {
+          if (n.nodeType === Node.ELEMENT_NODE && n.localName === 'style') {
+            const el = /** @type {!HTMLStyleElement} */(n);
+            if (this.isDocumentStyle(el)) {
+              this.addCustomStyle(el);
```
Add a comment here noting that `addCustomStyle` dedups so we don't have to worry about add/remove/add before validation.

azakus replied:
Yep, will do!

sorvell commented on src/document-wait.js:
```diff
@@ -10,36 +10,29 @@ subject to an additional IP rights grant found at http://polymer.github.io/PATEN
- */
-export default function documentWait(callback) {
-  requestAnimationFrame(function() {
+/** @const {Promise<void>} */
```
These changes are just simplified factoring?

azakus replied:
Yep, just a cleaner refactoring.

dfreedm committed:
Fix up tests and respond to feedback

Remove references to the no-longer-existing CustomStyleInterface build
artifact.

Clean up ApplyShim and ScopingShim use of CustomStyleInterface

Make sure all tests are passing

Make sure a late arrival of ApplyShim resets previously encountered
stylesheets in the document

dfreedm committed:
remove accidental pretty printing in closure

dfreedm committed:
fix closure warnings

dfreedm committed:
properly extern _resetCachedStyles on CustomStyleInterface

TimvdLippe reviewed (approved):
This seems like a good change to me! @azakus I assume this is a breaking change, waiting for a 2.0 branch of `ShadyCSS`?
